qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,141 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Cancel the redundant (self-adjoint) gates through commutation relations."""
+
+from collections import defaultdict
+
+import numpy as np
+
+from qiskit.circuit.classicalregister import ClassicalRegister
+from qiskit.circuit.commutation_checker import CommutationChecker
+from qiskit.circuit.commutation_library import standard_gates_commutations
+from qiskit.circuit.controlflow import CONTROL_FLOW_OP_NAMES
+from qiskit.circuit.instruction import Instruction
+from qiskit.circuit.library.generalized_gates.pauli import PauliGate
+from qiskit.circuit.library.standard_gates.p import PhaseGate
+from qiskit.circuit.library.standard_gates.rx import RXGate
+from qiskit.circuit.library.standard_gates.rz import RZGate
+from qiskit.circuit.library.standard_gates.u1 import U1Gate
+from qiskit.circuit.quantumregister import QuantumRegister, Qubit
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.dagcircuit import DAGCircuit, DAGInNode, DAGNode, DAGOutNode
+from qiskit.quantum_info import Pauli
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes.optimization.commutation_analysis import \
+    CommutationAnalysis
+from qiskit.transpiler.passmanager import PassManager
+
+commutator = CommutationChecker(standard_gates_commutations)
+
+
+class LightCone(TransformationPass):
+    """Remove the gates that do not affect the outcome of a measurement on a circuit.
+
+    Pass for computing the light-cone of an observable or measurement. The Pass can
+    handle either a Pauli operator one would like to measure or a measurement on a set
+    of qubits.
+    """
+
+    def __init__(self, observable: Pauli | None = None):
```

Comment:
This also takes a list of Paulis, right?

Reply:
For now only one Pauli. We could easily adapt it to a list of Paulis if we think that would make more sense.

Reply:
I'm asking because in your code the default path for `observable=None` seems to use a list of Paulis already 🙂

Reply:
A list of Pauli terms would, in general, lead to a collection of different circuits as each Pauli term may have its own independent causal structure.  A one circuit in, many circuits out, is not currently supported in the transpiler interface.  But perhaps should be as it is useful for twirling as well

Reply:
Oh I meant that in this case the gates would have to commute with each of the Paulis in the list (how the measurements are treated as list of single-Z Paulis in the code right now). That would e.g. allow to decompose a larger observable into smaller (e.g. `ZZZ...` into `Z0, Z1, ...`) -- this would give a larger light cone since now things might not commute which would in the bigger observable, but it would make the commutation check faster (which I think would use a matrix multiplication check for non-standard gates).

But that's probably not required for the beginning and we could just issue a warning if the commutation check becomes problematic 🙂